source/unix/openssl_rsa.c

```diff
@@ -41,53 +41,76 @@ static void s_rsa_destroy_key(void *key_pair) {
 }
 
 /*
-* Transforms evp error code into crt error code and raises it as necessary.
-* All evp functions follow the same:
-* >= 1 for success
-* <= 0 for failure
-* -2 always indicates incorrect algo for operation
-*/
+ * Transforms evp error code into crt error code and raises it as necessary.
+ * All evp functions follow the same:
+ * >= 1 for success
+ * <= 0 for failure
+ * -2 always indicates incorrect algo for operation
+ */
 static int s_reinterpret_evp_error_as_crt(int evp_error, const char *function_name) {
     if (evp_error > 0) {
         return AWS_OP_SUCCESS;
-    } 
+    }
 
+    /* AWS-LC/BoringSSL error code is uint32_t, but OpenSSL uses unsigned long. */
+#if !defined(OPENSSL_IS_AWSLC) && !defined(OPENSSL_IS_BORINGSSL)
     uint32_t error = ERR_peek_error();
+#else
+    unsigned long error = ERR_peek_error();
+#endif
+
+    char error_message[120] = {0};
+    int crt_error = AWS_OP_ERR;
+
     if (evp_error == -2) {
-        return aws_raise_error(AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM);
+        crt_error = AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM;
+        goto on_error;
     }
 
-    AWS_LOGF_ERROR(AWS_LS_CAL_RSA, "Calling function %s failed with %d and extended error %lu",
-        function_name, evp_error, (unsigned long)error);
-
     if (ERR_GET_LIB(error) == ERR_LIB_EVP) {
         switch (ERR_GET_REASON(error)) {
-            case EVP_R_BUFFER_TOO_SMALL:
-                return aws_raise_error(AWS_ERROR_SHORT_BUFFER);
-            case EVP_R_UNSUPPORTED_ALGORITHM:
-                return aws_raise_error(AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM);
+            case EVP_R_BUFFER_TOO_SMALL: {
+                crt_error = AWS_ERROR_SHORT_BUFFER;
+                goto on_error;
+            }
+            case EVP_R_UNSUPPORTED_ALGORITHM: {
+                crt_error = AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM;
+                goto on_error;
+            }
         }
     }
 
-    return aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);   
+    crt_error = AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED;
+
+on_error:
+    ERR_error_string_n(error, error_message, sizeof(error_message));
+    AWS_LOGF_ERROR(
+        AWS_LS_CAL_RSA,
+        "%s() failed. returned: %d extended error:%lu(%s) aws_error:%s",
```

Comment:
ERR_error_string_n() writes out `error:[error code]:[library name]::[reason string]`, so you don't need to print the `%lu`.

If you just wanted the reason string, you can call `const char *ERR_reason_error_string(e)` without needing an intermediate buffer

Also, you're calling it even when `error == 0`, and I'm not sure what it prints in that case.

Reply:
ERR_reason_error_string was not showing up in intellisense, when it tried it, so i guessed it was not available in lc, but looks like thats wrong. switched to it now. 
for error error code, NULL is returned so it needs to guard against it